Mathlib/Analysis/NormedSpace/PiTensorProduct/InjectiveSeminorm.lean

```diff
@@ -0,0 +1,467 @@
+
+open scoped TensorProduct
+
+open BigOperators
+
+namespace PiTensorProduct
+
+section seminorm
+
+lemma toDualMultilinearMap_bound (x : ⨂[𝕜] i, E i) :
+    ∃ (C : ℝ), 0 ≤ C ∧ ∀ (G : Type*) [SeminormedAddCommGroup G]
+    [NormedSpace 𝕜 G] (f : ContinuousMultilinearMap 𝕜 E G),
+    ‖lift f.toMultilinearMap x‖ ≤ C * ‖f‖ := by
+  induction' x using PiTensorProduct.induction_on with r m x y hx hy
+  · existsi ‖r‖ * ∏ i : ι, ‖m i‖
+    constructor
+    · exact mul_nonneg (norm_nonneg r) (Finset.prod_nonneg (fun i _ ↦ norm_nonneg (m i)))
+    · intro G _ _ f
+      simp only [map_smul, lift.tprod, ContinuousMultilinearMap.coe_coe]
+      rw [mul_assoc, mul_comm _ ‖f‖, norm_smul]
+      exact le_trans (mul_le_mul_of_nonneg_left (ContinuousMultilinearMap.le_opNorm f m)
+        (norm_nonneg r)) (le_refl _)
+  · obtain ⟨Cx, hCx⟩ := hx; obtain ⟨Cy, hCy⟩ := hy
+    existsi Cx + Cy
+    constructor
+    · exact add_nonneg hCx.1 hCy.1
+    · intro G _ _ f
+      rw [map_add, add_mul]
+      refine le_trans (norm_add_le _ _) (add_le_add (hCx.2 _ f) (hCy.2 _ f))
+
+/-- The linear map from `⨂[𝕜] i, Eᵢ` to `ContinuousMultilinearMap 𝕜 E F →L[𝕜] F` sending
+`x` in `⨂[𝕜] i, Eᵢ` to the map `f ↦ f.lift x`.
+-/
+@[simps!]
+noncomputable def toDualContinuousMultilinearMap : (⨂[𝕜] i, E i) →ₗ[𝕜]
+    ContinuousMultilinearMap 𝕜 E F →L[𝕜] F where
+  toFun x := LinearMap.mkContinuousOfExistsBound
+    ((LinearMap.flip (lift (R := 𝕜) (s := E) (E := F)).toLinearMap x) ∘ₗ
+    ContinuousMultilinearMap.toMultilinearMapLinear) (by
+      obtain ⟨C, hC⟩ := toDualMultilinearMap_bound x
+      existsi C
+      intro f
+      simp only [LinearMap.coe_comp, Function.comp_apply,
+        ContinuousMultilinearMap.toMultilinearMapLinear_apply, LinearMap.flip_apply,
+        LinearEquiv.coe_coe]
+      exact hC.2 F f)
+  map_add' x y := by
+    ext _
+    simp only [map_add, LinearMap.mkContinuousOfExistsBound_apply, LinearMap.coe_comp,
+      Function.comp_apply, ContinuousMultilinearMap.toMultilinearMapLinear_apply,
+      LinearMap.add_apply, LinearMap.flip_apply, LinearEquiv.coe_coe, ContinuousLinearMap.add_apply]
+  map_smul' a x := by
+    ext _
+    simp only [map_smul, LinearMap.mkContinuousOfExistsBound_apply, LinearMap.coe_comp,
+      Function.comp_apply, ContinuousMultilinearMap.toMultilinearMapLinear_apply,
+      LinearMap.smul_apply, LinearMap.flip_apply, LinearEquiv.coe_coe, RingHom.id_apply,
+      ContinuousLinearMap.coe_smul', Pi.smul_apply]
+
+/-- The injective seminorm on `⨂[𝕜] i, Eᵢ`. Morally, it sends `x` in `⨂[𝕜] i, Eᵢ` to the
+`sup` of the operator norms of the `PiTensorProduct.toDualContinuousMultilinearMap x`, for all
+normed vector spaces `F`. In fact, we only take in the same universe as `⨂[𝕜] i, Eᵢ`, and then
+prove in `PiTensorProduct.injectiveSeminorm_bound` that this gives the same result.
+-/
+noncomputable def injectiveSeminorm : Seminorm 𝕜 (⨂[𝕜] i, E i) :=
+  sSup {p | ∃ (G : Type (max (max u_1 u_2) u_3)) (_ : SeminormedAddCommGroup G)
```

Comment:
Here it's impossible to guess what the universes are (and if someone adds a new `Type*` it will break everything). Better use explicit universe variables in 𝕜, E, F (naming them for instance `u𝕜`, `uE`, `uF`).

Reply:
Good point, done.